antlr-kotlin-runtime/src/commonMain/kotlin/com/strumenta/antlrkotlin/runtime/ext/String.ext.kt

```diff
@@ -2,15 +2,48 @@
 // Copyright 2024-present Strumenta and contributors, licensed under BSD 3-Clause.
 package com.strumenta.antlrkotlin.runtime.ext
 
+import kotlin.math.min
+
+/**
+ * Returns the code points in this string.
+ */
 public fun String.codePointIndices(): IntArray {
-  val intArray = IntArray(length)
-  var p = 0
+  val (array, size) = codePointIndicesFast(this)
+  return array.copyOfRange(0, size)
+}
+
+/**
+ * Returns the code points in [str], and the real size
+ * of the returned array, since it is not trimmed to avoid
+ * an additional potentially expensive allocation.
+ */
+internal fun codePointIndicesFast(str: String): Pair<IntArray, Int> {
+  val strLength = str.length
+  val intArray = IntArray(strLength + 1)
+  var size = 0
+  var i = 1
 
-  for (i in indices) {
-    if (!hasSurrogatePairAt(i - 1)) {
-      intArray[p++] = i
+  intArray[size++] = 0
+
+  while (i < strLength) {
+    if (!hasSurrogatePairAtFast(str, i)) {
+      intArray[size++] = i
     }
+
+    i++
+  }
+
+  return Pair(intArray, min(size, strLength))
+}
+
+/**
+ * Returns whether [str] has a unicode surrogate pair at the specified [index].
+ */
+private fun hasSurrogatePairAtFast(str: String, index: Int): Boolean {
+  // Note(Edoardo): keep it like this for performance reasons
+  if (str[index - 1].code in /* MIN_HIGH_SURROGATE */ 0xD800..0xDBFF /* MAX_HIGH_SURROGATE */) {
```

Comment:
I think the comment makes sense to avoid the temptation to refactor the code.

Are you also saying that replacing those values with references to the corresponding constants slow down the code? If so, is there a way to avoid that (is there a sort of inline for constants?)
I would find this quite surprising

Reply:
> is there a sort of inline for constants

A `const val` is normally inlined, but it does not work in JS when that `const val` is inside of a `companion object`, since it is lazily initialized.

Without proper inlining you'll have a null check (two calls deep) for every character in the string, losing significant time.

Reply:
I would then put this explanation in a comment in the code

Reply:
Added.